#+TITLE: GBS samples
#+PROPERTY:  header-args :var DIR=(file-name-directory (file-local-name buffer-file-name))

* Dried Tissue

| IGBB sample ID | Species            | Line ID          |
|----------------+--------------------+------------------|
| Kd-3041        | Kokia drynarioides | HA-PWM-A_3041    |
| Kd-3042        | Kokia drynarioides | HA-PWM-A_3042    |
| Kd-3043        | Kokia drynarioides | HA-PWM-A_3043    |
| Kc-F1          | Kokia cookei       | MO-LCC-F1        |
| Kc-0001        | Kokia cookei       | MA-AH1-A-0001    |
| Kc-01/01       | Kokia cookei       | MA-AH1-A # 01/01 |
| Kc-02          | Kokia cookei       | MA-AH1-A # 02    |
| Kc-03          | Kokia cookei       | MA-AH1-A # 03    |
| Kc-04          | Kokia cookei       | MA-AH1-A # 04    |
| Kc-05          | Kokia cookei       | MA-AH1-A # 05    |
| Kc-06          | Kokia cookei       | MA-AH1-A # 06    |
| Kc-07          | Kokia cookei       | MA-AH1-A # 07    |
| Kc-08          | Kokia cookei       | MA-AH1-A # 08    |
| Kc-09          | Kokia cookei       | MA-AH1-A # 09    |
| Kc-10          | Kokia cookei       | MA-AH1-A # 10    |
| Kc-11          | Kokia cookei       | MA-AH1-A # 11    |

* DNA

|         Sample ID | Species                   |
|-------------------+---------------------------|
|              3148 | Kokia cookei              |
|              3149 | Kokia drynarioides        |
|              3150 | Kokia drynarioides        |
|              3151 | Kokia kauaiensis          |
|              3152 | Kokia kauaiensis          |
|              3153 | Kokia kauaiensis          |
|              3154 | Kokia kauaiensis          |
|              3155 | Kokia kauaiensis          |
|              3162 | Kokia kauaiensis          |
|              3163 | Kokia kauaiensis          |
|              3156 | Kokia kauaiensis          |
|              3157 | Kokia kauaiensis          |
|              3158 | Kokia kauaiensis          |
|              3159 | Kokia kauaiensis          |
|              3160 | Kokia kauaiensis          |
|            3160_2 | Kokia kauaiensis          |
|              3161 | Kokia kauaiensis          |
|              3162 | Kokia kauaiensis          |
|              3163 | Kokia kauaiensis          |
|              3164 | Kokia kauaiensis          |
|              3165 | Kokia kauaiensis          |
|              3166 | Kokia kauaiensis          |
|              3167 | Kokia kauaiensis          |
|              3168 | Kokia kauaiensis          |
| 3160_re-extracted | Kokia kauaiensis          |
|              3169 | Kokia kauaiensis          |
|              3170 | Kokia kauaiensis          |
|              3171 | Kokia kauaiensis          |
|              3172 | Kokia kauaiensis          |
|              3172 | Kokia kauaiensis          |
|              3174 | Kokia kauaiensis          |
|              3175 | Kokia kauaiensis          |
|              3176 | Kokia kauaiensis          |
|   3164_re_extract | Kokia kauaiensis          |
|              3177 | Kokia kauaiensis          |
|              3178 | Kokia kauaiensis          |
|              3179 | Kokia kauaiensis          |
|              3180 | Kokia kauaiensis          |
|              3181 | Kokia kauaiensis          |
|              3182 | Kokia kauaiensis          |
|              3183 | Kokia kauaiensis          |
|              3184 | Kokia kauaiensis          |
|   3169_re-extract | Kokia kauaiensis          |
|              3185 | Kokia kauaiensis          |
|              3186 | Kokia kauaiensis          |
|              3187 | Kokia kauaiensis          |
|              3188 | Kokia kauaiensis          |
|              3189 | Kokia kauaiensis          |
|              3190 | Kokia kauaiensis          |
|              3191 | Kokia kauaiensis          |
|              3192 | Kokia kauaiensis          |
|   3170_re-extract | Kokia kauaiensis          |
|              3193 | Kokia kauaiensis          |
|              3194 | Kokia kauaiensis          |
|              3195 | Kokia kauaiensis          |
|              3215 | Kokia kauaiensis          |
|              8046 | Kokia sp. cf.drynarioides |
|              3330 | Kokia kauaiensis          |
|              3331 | Kokia kauaiensis          |
|              3332 | Kokia kauaiensis          |
|              3333 | Kokia kauaiensis          |
|               283 | Kokia drynarioides        |
|              1780 | Kokia drynarioides        |
|              4372 | Kokia cookei              |
|              4373 | Kokia cookei              |



* Align

#+NAME: kd-gbs
| Name       | Forward                                                 | Reverse                                                 |
|------------+---------------------------------------------------------+---------------------------------------------------------|
| Kd_3041    | Kd_3041/Kd_3041_CKDL220020121-1A_HCWYNDSX5_L1_1.fq.gz   | Kd_3041/Kd_3041_CKDL220020121-1A_HCWYNDSX5_L1_2.fq.gz   |
| Kd_3042    | Kd_3042/Kd_3042_CKDL220020121-1A_HCWYNDSX5_L1_1.fq.gz   | Kd_3042/Kd_3042_CKDL220020121-1A_HCWYNDSX5_L1_2.fq.gz   |
| Kd_3043    | Kd_3043/Kd_3043_CKDL220020121-1A_HCWYNDSX5_L1_1.fq.gz   | Kd_3043/Kd_3043_CKDL220020121-1A_HCWYNDSX5_L1_2.fq.gz   |
| Kd_JW      | Kd_JW/Kd_JW_CKDL220020121-1A_HCWYNDSX5_L1_1.fq.gz       | Kd_JW/Kd_JW_CKDL220020121-1A_HCWYNDSX5_L1_2.fq.gz       |
| Kd_JW_ng   | Kd_JW_ng/Kd_JW_ng_CKDL220020121-1A_HCWYNDSX5_L1_1.fq.gz | Kd_JW_ng/Kd_JW_ng_CKDL220020121-1A_HCWYNDSX5_L1_2.fq.gz |
| SRR6195040 | SRR6195040/SRR6195040_1.fastq.gz                        | SRR6195040/SRR6195040_2.fastq.gz                        |
| Kk_19S8    | Kk_19S8/Kk_19S8_CKDL220020121-1A_HCWYNDSX5_L1_1.fq.gz   | Kk_19S8/Kk_19S8_CKDL220020121-1A_HCWYNDSX5_L1_2.fq.gz   |

- Create bwa database for assembly
  #+begin_src sh :tangle 1-align/0-ref.kd.sh
ROOT=$(git rev-parse --show-toplevel)

ml singularity
bwa () {
    singularity exec -B $ROOT \
        /apps/singularity-3/bwa/bwa-0.7.17.sif bwa $@
}
samtools () {
    singularity exec -B $ROOT \
        /apps/singularity-3/samtools/samtools-v1.9-4-deb_cv1.sif \
        samtools $@
}
fasta=$DIR/0-ref/Kd.contigs.fa
bwa index -a bwtsw $fasta
samtools faidx $fasta
  #+end_src
- Align to contigs
  #+header: :var data=kd-gbs[8:9,]
  #+begin_src sh :tangle 1-align/1-bwa.sh
ROOT=$(git rev-parse --show-toplevel)
RAW=$(realpath $DIR/0-raw/)

ml singularity
bwa () {
    singularity exec -B $ROOT \
        /apps/singularity-3/bwa/bwa-0.7.17.sif bwa $@
}
samtools () {
    singularity exec -B $ROOT \
        /apps/singularity-3/samtools/samtools-v1.9-4-deb_cv1.sif \
        samtools $@
}
PATH=$PATH:$ROOT/wgs/1-assembly/apps/samblaster-v.0.1.26/

fasta=$DIR/0-ref/Kd.contigs.fa
for name in "${!data[@]}"; do
    readarray -t libs <<< "${data[$name]}"

    bwa mem -t 48 $fasta $RAW/${libs[0]} $RAW/${libs[1]} |
        samblaster |
        samtools view -bS |
        samtools sort -m200G -o $DIR/1-align/$name.bam -
done

  #+end_src

* Coverage
- Calculate Base Depth
  #+header: :var data=kd-gbs[8,0]
  #+begin_src sh :tangle 2-coverage/1-depth.sh
ROOT=$(git rev-parse --show-toplevel)
export DIR ROOT

ml singularity
samtools () {
    singularity exec -B $ROOT \
        /apps/singularity-3/samtools/samtools-v1.9-4-deb_cv1.sif \
        samtools $@
}
export -f samtools

parallel --eta samtools depth $DIR/1-align/{}.bam \
    '>' $DIR/2-coverage/{}.depth ::: "${data[@]}"

  #+end_src

* Variant calling

#+header: :var data=kd-gbs[,0]
#+begin_src sh :tangle 3-variants/1-call.sh
ROOT=$(git rev-parse --show-toplevel)
RAW=$(realpath $DIR/0-raw/)

ml singularity
bcftools () {
    singularity exec -B $ROOT \
        /apps/singularity-3/bcftools/bcftools-1.9.sif \
        bcftools "$@"
}

fasta=$DIR/0-ref/Kd.contigs.fa

bcftools mpileup -Ou -f $fasta $(printf "$DIR/1-align/%s.bam " "${data[@]}") |
    bcftools call -vmO z -o 3-variants/all.vcf.gz

#+end_src

#+NAME: plot-pca
#+HEADER: :results output graphics file  :width 800 :height 600 :noweb yes
#+begin_src R :file kd-samples.png :var vcf='3-variants/kd.vcf'
library(tidyverse)
library(SNPRelate)
library(SeqArray)
library(ggrepel)

vcf <- scan(text=vcf, what="character", allowEscapes = T)

# Using SeqArray converter instead of SNPRelate's
seqVCF2GDS(vcf, 'all.gds')
genofile <- seqOpen('all.gds')
# Run PCA using SNPRelate
#snpgdsVCF2GDS(vcf, 'all.gds', method='biallelic.only')
#genofile <- snpgdsOpen('all.gds')
pca <- snpgdsPCA(genofile, autosome.only=F, num.thread=48)
pc.percent <- pca$varprop*100

## Grab relevent data
data <- data.frame(Library.ID = pca$sample.id,
    EV1 = pca$eigenvect[,1],    # the first eigenvector
    EV2 = pca$eigenvect[,2],    # the second eigenvector
    stringsAsFactors = FALSE) %>%
  mutate(label = basename(Library.ID))

# Plot PCA, labeling outliers
p <- ggplot(data, aes(x=EV1, y=EV2)) +
  geom_point() +
  geom_label_repel(aes(label=label)) +
  scale_shape_manual(values=c(20, 4)) +
  xlab(sprintf("PC 1 (%0.2f%%)", pc.percent[1])) +
  ylab(sprintf("PC 2 (%0.2f%%)", pc.percent[2])) +
  theme_minimal()
p
#+end_src

[[file:all-samples.png]]
