#+TITLE: Hi-C Scaffolding
#+PROPERTY:  header-args :var DIR=(file-name-directory buffer-file-name)

#+BEGIN_SRC sh :tangle 0-genome/setup.sh
cd $DIR
ROOT=$(git rev-parse --show-toplevel)

ml igbb bwa

cat $ROOT/assemble/1-raw/canu/kokia.contigs.fasta >  $DIR/0-genome/kokia.fa

bwa index $DIR/0-genome/kokia.fa
python $DIR/bin/juicer/misc/generate_site_positions.py MboI $DIR/0-genome/kokia{,.fa}
awk '{print $1, $NF}' OFS="\t" $DIR/0-genome/kokia_MboI.txt > $DIR/0-genome/kokia.sizes
#+END_SRC

* 3D-DNA

#+BEGIN_SRC sh :tangle 1-juicer/setup.sh
cd $DIR
ROOT=$(git rev-parse --show-toplevel)

mkdir $DIR/1-juicer/fastq/
ln -s $ROOT/raw/hi-c/kokia_S3HiC_R1.fastq.gz $DIR/1-juicer/fastq/Kokia_HiC_R1.fastq.gz
ln -s $ROOT/raw/hi-c/kokia_S3HiC_R2.fastq.gz $DIR/1-juicer/fastq/Kokia_HiC_R2.fastq.gz
ln -s $DIR/bin/juicer/CPU bin/scripts
#+END_SRC

#+BEGIN_SRC sh :tangle 1-juicer/run.sh
cd $DIR
ROOT=$(git rev-parse --show-toplevel)

ml igbb bwa

$DIR/bin/scripts/juicer.sh \
    -p $DIR/0-genome/kokia.sizes \
    -y $DIR/0-genome/kokia_MboI.txt \
    -z $DIR/0-genome/kokia.fa \
    -t $PBS_NUM_PPN \
    -D $DIR/bin/ \
    -d $DIR/1-juicer/
#+END_SRC


#+BEGIN_SRC sh :tangle  2-3d-dna/run.sh
cd $DIR
ROOT=$(git rev-parse --show-toplevel)

unset PERL5LIB
ml igbb bwa

PATH=$DIR/bin/3d-dna/bin:$PATH
PATH=$DIR/bin/lastz-1.04.00/src:$PATH
PATH=/usr/local/igbb/coreutils-8.24/bin:$PATH

cd $DIR/2-3d-dna
$DIR/bin/3d-dna/run-asm-pipeline.sh \
    -m diploid \
    $DIR/0-genome/kokia.fa  \
    $DIR/1-juicer/aligned/merged_nodups.txt



#+END_SRC

#+BEGIN_SRC sh
/usr/local/igbb/abyss-2.0.1/bin/abyss-fac $DIR/2-3d-dna/kokia.FINAL.fasta
#+END_SRC

#+RESULTS:
|    n | n:500 | L50 |  min |     N80 |     N50 |     N20 |  E-size |     max |     sum |
|------+-------+-----+------+---------+---------+---------+---------+---------+---------|
| 2569 |  2569 |  16 | 1000 | 4912341 | 11.84e6 | 19.43e6 | 11.51e6 | 24.02e6 | 511.1e6 |

*  Salsa

#+BEGIN_SRC sh :tangle 1-bwa/run.sh
cd $DIR
ROOT=$(git rev-parse --show-toplevel)

ml igbb bwa samtools
source $DIR/bin/slurm-hic/bin/activate
PATH=$PATH:$DIR/bin/slurm-hic/

bwa mem -t $PBS_NUM_PPN $DIR/0-genome/kokia.fa $ROOT/raw/hi-c/kokia_S3HiC_R1.fastq.gz |
    samtools view -bh - |
    filter_chimeras.py - > $DIR/1-bwa/r1.bam
bwa mem -t $PBS_NUM_PPN $DIR/0-genome/kokia.fa $ROOT/raw/hi-c/kokia_S3HiC_R2.fastq.gz |
    samtools view -bh - |
    filter_chimeras.py - > $DIR/1-bwa/r2.bam

#+END_SRC

#+BEGIN_SRC sh :tangle 1-bwa/combine.sh
cd $DIR
ROOT=$(git rev-parse --show-toplevel)

ml igbb bwa samtools
source $DIR/bin/slurm-hic/bin/activate
PATH=$PATH:$DIR/bin/slurm-hic/

combine_ends.py $DIR/1-bwa/r1.bam $DIR/1-bwa/r2.bam |
    samtools fixmate -m - - |
    samtools sort - |
    samtools markdup -r - $DIR/1-bwa/combined.bam

#+END_SRC

#+BEGIN_SRC sh :tangle 2-salsa/setup.sh
cd $DIR
ROOT=$(git rev-parse --show-toplevel)

ml igbb samtools
ml singularity

singularity exec -B$ROOT $DIR/bin/bedtools-2.27.sif \
    bamToBed -i $DIR/1-bwa/combined.bam |
    sort -k 4 > $DIR/2-salsa/combined.bed
#+END_SRC

#+BEGIN_SRC sh :tangle 2-salsa/run.sh
cd $DIR
ROOT=$(git rev-parse --show-toplevel)

ml python/2.7.15
source $DIR/bin/SALSA-2.2/bin/activate
mkdir $DIR/2-salsa/scaffolds

python2 $DIR/bin/SALSA-2.2/run_pipeline.py \
    -a $DIR/0-genome/kokia.fa \
    -l $DIR/0-genome/kokia.fa.fai \
    -b $DIR/2-salsa/combined.bed \
    -e GATC \
    -o $DIR/2-salsa/scaffolds-no-mis \
    -p yes
#+END_SRC

#+BEGIN_SRC sh
/usr/local/igbb/abyss-2.0.1/bin/abyss-fac $DIR/2-salsa/scaffolds-no-mis/scaffolds_FINAL.fasta
#+END_SRC

#+RESULTS:
|    n | n:500 | L50 |  min |     N80 |     N50 |     N20 |  E-size |     max |     sum |
|------+-------+-----+------+---------+---------+---------+---------+---------+---------|
| 1166 |  1166 |  24 | 1020 | 2741557 | 6467607 | 12.31e6 | 8302211 | 24.26e6 | 532.8e6 |
