#+TITLE: Assembly
#+PROPERTY:  header-args :var DIR=(file-name-directory buffer-file-name)

* Kokia drynarioides JFW-HI

- Remove lambda and remove reads shorter than 4kb from nanopore data
  #+header: :var runs=../0-raw/readme.org:kd-nanopore-runs[,1]
  #+begin_src sh :tangle kd/1-trim.sh
RAW=$(realpath $DIR/../0-raw/)

printf "$RAW/%s/fastq_pass.fq.gz\n" "${runs[@]}" |
    xargs $DIR/apps/devour -t 20 -l 4000 |
    gzip > kd/clean.fq.gz
  #+end_src

- Assemble contigs using canu
  #+begin_src sh :tangle kd/2-canu.sh
ml canu/2.1
canu \
    -p canu -d $DIR/kd \
    genomeSize=590m \
    gridOptions="-A191400-327070 --time=48:00:00" \
    useGrid=true \
    -nanopore $DIR/kd/clean.fq.gz


  #+end_src

|        | canu/2.1 |
|--------+----------|
| n      |     2264 |
| n:500  |     2264 |
| L50    |       54 |
| min    |     1197 |
| N75    |  1201645 |
| N50    |  2757100 |
| N25    |  4885516 |
| E-size |  4190213 |
| max    |  22.95e6 |
| sum    |  551.1e6 |

- clean canu directory
  #+begin_src sh
pushd $DIR/kd
rm -fr canu-logs/ \
    canu-scripts/ \
    canu.seqStore/ \
    correction/ \
    trimming/ \
    unitigging/ \
    canu.out

gzip *.fasta canu.contigs.layout.*

git add canu.report canu.seqStore.err

git annex add canu.contigs.fasta.gz \
    canu.contigs.layout.readToTig.gz \
    canu.contigs.layout.tigInfo.gz \
    canu.correctedReads.fasta.gz \
    canu.trimmedReads.fasta.gz \
    canu.unassembled.fasta.gz

#+end_src
** sup model

Using the 'Nanopore flip-flop R9.4 or R10.3' settings because of the reduced
error rate for the latest basecalling data.

#+begin_src sh :tangle kd.sup/run.sh
RAW=$(realpath $DIR/../0-raw/)
PATH=$DIR/apps/canu-2.2/bin:$PATH

canu \
    -p canu -d $DIR/kd.sup \
    genomeSize=590m \
    gridOptions="-Acotton_genomics" \
    useGrid=true \
    corMhapOptions='--threshold 0.8 --ordered-sketch-size 1000 --ordered-kmer-size 14' \
    correctedErrorRate=0.105 \
    -nanopore $RAW/*/sup.pass.fq.gz


#+end_src

| n      |               2365 |
| n:500  |               2365 |
| L50    |                 89 |
| min    |               1206 |
| N75    |             755687 |
| N50    |            1659488 |
| N25    |            3113363 |
| E-size |            2550534 |
| max    |            11.36e6 |
| sum    |            552.8e6 |
| name   | canu.contigs.fasta |

- Clean canu directory (see previous)

* Kokia kauaiensis S9

- 1st run
  #+begin_src sh :tangle kk/run.sh
RAW=$(realpath $DIR/../0-raw/)
PATH=$DIR/apps/canu-2.2/bin:$PATH

canu \
    -p canu -d $DIR/kk \
    genomeSize=590m \
    gridOptions="-Acotton_genomics" \
    useGrid=true \
    -nanopore $RAW/Kk/*/fastq_pass.fq.gz


#+end_src
| n      |                 16092 |
| n:500  |                 16092 |
| L50    |                  2530 |
| min    |                  1114 |
| N75    |                 27190 |
| N50    |                 56809 |
| N25    |                108701 |
| E-size |                 84051 |
| max    |                641467 |
| sum    |               536.3e6 |
| name   | kk/canu.contigs.fasta |


* Kokia cookei C69

- 1st run
  #+begin_src sh :tangle kc/run.sh
RAW=$(realpath $DIR/../0-raw/)
PATH=$DIR/apps/canu-2.2/bin:$PATH

canu \
    -p canu -d $DIR/kc \
    genomeSize=590m \
    gridOptions="-Acotton_genomics" \
    useGrid=true \
    -nanopore $RAW/Kc/*/fastq_pass.fq.gz


#+end_src
| n      |                  4992 |
| n:500  |                  4992 |
| L50    |                   414 |
| min    |                  1027 |
| N75    |                147619 |
| N50    |                342606 |
| N25    |                702028 |
| E-size |                561401 |
| max    |               4034082 |
| sum    |               555.8e6 |
| name   | kc/canu.contigs.fasta |
