#+TITLE: Assembly
#+PROPERTY:  header-args :var DIR=(file-name-directory buffer-file-name)

* Kokia drynarioides JFW-HI

** Contigs
- Remove lambda and remove reads shorter than 4kb from nanopore data
  #+header: :var runs=../0-raw/readme.org:kd-nanopore-runs[,1]
  #+begin_src sh :tangle kd/1-trim.sh
RAW=$(realpath $DIR/../0-raw/)

printf "$RAW/%s/fastq_pass.fq.gz\n" "${runs[@]}" |
    xargs $DIR/apps/devour -t 20 -l 4000 |
    gzip > kd/clean.fq.gz
  #+end_src

- Assemble contigs using canu
  #+begin_src sh :tangle kd/2-canu.sh
ml canu/2.1
canu \
    -p canu -d $DIR/kd \
    genomeSize=590m \
    gridOptions="-A191400-327070 --time=48:00:00" \
    useGrid=true \
    -nanopore $DIR/kd/clean.fq.gz


  #+end_src

|        | canu/2.1 |
|--------+----------|
| n      |     2264 |
| n:500  |     2264 |
| L50    |       54 |
| min    |     1197 |
| N75    |  1201645 |
| N50    |  2757100 |
| N25    |  4885516 |
| E-size |  4190213 |
| max    |  22.95e6 |
| sum    |  551.1e6 |

- clean canu directory
  #+begin_src sh
pushd $DIR/kd
rm -fr canu-logs/ \
    canu-scripts/ \
    canu.seqStore/ \
    correction/ \
    trimming/ \
    unitigging/ \
    canu.out

gzip *.fasta canu.contigs.layout.*

git add canu.report canu.seqStore.err

git annex add canu.contigs.fasta.gz \
    canu.contigs.layout.readToTig.gz \
    canu.contigs.layout.tigInfo.gz \
    canu.correctedReads.fasta.gz \
    canu.trimmedReads.fasta.gz \
    canu.unassembled.fasta.gz

#+end_src

- dotplot against kirkii

  #+begin_src sh
ml minimap2/2.17 r/4.0.25

minimap2 -x asm5 -t 20 \
    $DIR/0-ref/kirkii.fa $DIR/kd/canu.contigs.fasta \
    > $DIR/kd/canu.kirkii.paf

Rscript $DIR/apps/dotplots/pafCoordsDotPlotly.R -i $DIR/kd/canu.kirkii.paf \
    -o $DIR/kd/canu.kirkii -m 20000 -x -s -t -l -p 12
  #+end_src


** Hi-C Scaffolding
- Create bwa database for assembly
  #+begin_src sh :tangle kd/2-hic/0-ref/run.sh
ROOT=$(git rev-parse --show-toplevel)

ml singularity/3.7.3
bwa () {
    singularity exec -B $ROOT \
        /apps/singularity-3/bwa/bwa-0.7.17.sif bwa $@
}
samtools () {
    singularity exec -B $ROOT \
        /apps/singularity-3/samtools/samtools-v1.9-4-deb_cv1.sif \
        samtools $@
}

bwa index -a bwtsw -p $DIR/kd/2-hic/0-ref/kd.bwa $DIR/kd/canu.contigs.fasta
samtools faidx $DIR/kd/canu.contigs.fasta
  #+end_src
- Align Hi-C data to assembly
  #+header: :var data=../0-raw/readme.org:kd-hic
  #+begin_src sh :tangle kd/2-hic/1-bwa/run.sh
ROOT=$(git rev-parse --show-toplevel)
RAW=$(realpath $DIR/../0-raw/)

ml singularity/3.7.3
bwa () {
    singularity exec -B $ROOT \
        /apps/singularity-3/bwa/bwa-0.7.17.sif bwa $@
}
samtools () {
    singularity exec -B $ROOT \
        /apps/singularity-3/samtools/samtools-v1.9-4-deb_cv1.sif \
        samtools $@
}
PATH=$PATH:$DIR/apps/samblaster

for name in "${!data[@]}"; do
    readarray -t libs <<< "${data[$name]}"

    bwa mem -5SP -t 24 $DIR/kd/2-hic/0-ref/kd.bwa $RAW/${libs[0]} $RAW/${libs[1]} |
        samblaster |
        samtools view -bS -F 2316 > $DIR/kd/2-hic/1-bwa/$name.bam
done

  #+end_src

- Validate libraries
  #+header: :var data=../0-raw/readme.org:kd-hic
  #+begin_src sh :tangle 2-hic/2-qc/run.sh
ROOT=$(git rev-parse --show-toplevel)

ml python/3.9.2
export PYTHONPATH=$PYTHONPATH:~/.local/lib/python3.9/site-packages/
source $DIR/apps/hic_qc/bin/activate

for name in "${!data[@]}"; do
    $DIR/apps/hic_qc/hic_qc.py \
        -b $DIR/2-hic/1-bwa/$name.bam \
        -o $DIR/2-hic/2-qc/$name \
        --lib_enzyme GATC
done

  #+end_src

- Convert bams to single bed
  #+header: :var data=../0-raw/readme.org:kd-hic[,0]
  #+begin_src sh :tangle kd/2-hic/3-convert.sh
ROOT=$(git rev-parse --show-toplevel)

ml singularity/3.7.3
bedtools () {
    singularity exec -B $ROOT \
        /apps/singularity-3/bedtools/bedtools-2.28.0.sif bedtools $@
}

export ROOT DIR
export -f bedtools

parallel --eta bedtools bamtobed -i $DIR/kd/2-hic/1-bwa/{}.bam \
    ::: "${data[@]}" > $DIR/kd/2-hic/3-alignments.bed
  #+end_src
- Run SALSA
  #+begin_src sh :tangle kd/2-hic/4-salsa/run.sh
ROOT=$(git rev-parse --show-toplevel)

ml singularity/3.7.3

singularity exec -B $ROOT $DIR/apps/salsa2_2.3--py27h16ec135_1.sif \
    run_pipeline.py \
    -i 15 \
    -a $DIR/kd/canu.contigs.fasta \
    -l $DIR/kd/canu.contigs.fasta.fai \
    -b $DIR/kd/2-hic/3-alignments.bed \
    -e GATC \
    -s 590000000 \
    -o $DIR/kd/2-hic/4-salsa/ \
    -p yes
  #+end_src

  #+begin_src sh :tangle kd/2-hic/4-salsa/stat.sh
ROOT=$(git rev-parse --show-toplevel)

ml singularity/3.7.3

singularity exec -B $ROOT \
    /apps/singularity-3/abyss/abyss-v2.1.5-7-deb_cv1.sif \
    /usr/lib/abyss/abyss-fac \
    $DIR/kd/2-hic/4-salsa/scaffolds_ITERATION_*.fasta \
    $DIR/kd/2-hic/4-salsa/scaffolds_FINAL.fasta 
    

  #+end_src

  #+RESULTS:
  |    n | n:500 | L50 |  min |     N75 |     N50 |     N25 |  E-size |        max |         sum |  name |
  | 1832 |  1832 |  48 | 1197 | 1337726 | 3293652 | 5573229 | 4604924 | 22950000.0 | 551100000.0 |     1 |
  | 1663 |  1663 |  46 | 1197 | 1455819 | 3362451 | 6249823 | 4727382 | 22950000.0 | 551100000.0 |     2 |
  | 1596 |  1596 |  45 | 1197 | 1680764 | 3702737 | 6262271 | 4841897 | 22950000.0 | 551100000.0 |     3 |
  | 1571 |  1571 |  43 | 1197 | 1722624 | 3786187 | 6299587 | 4959276 | 22950000.0 | 551100000.0 |     4 |
  | 1596 |  1596 |  45 | 1197 | 1680764 | 3702737 | 6262271 | 4841897 | 22950000.0 | 551100000.0 | FINAL |


- dotplot against kirkii

  #+begin_src sh
ml minimap2/2.17 r/4.0.25

cd $DIR/kd/

minimap2 -x asm5 -t 20 \
    $DIR/0-ref/kirkii.fa $DIR/kd/2-hic/4-salsa/scaffolds_ITERATION_4.fasta \
    > $DIR/kd/hic.kirkii.paf

Rscript $DIR/apps/dotplots/pafCoordsDotPlotly.R -i $DIR/kd/hic.kirkii.paf \
    -o hic.kirkii -m 20000 -x -s -t -l -p 12
  #+end_src
** Arrange/Orient
- Orient and arrange using /Gossipiodes kirkii/
  #+begin_src sh :tangle kd/3-ragtag/run.sh
ROOT=$(git rev-parse --show-toplevel)
ml singularity

ragtag () {
    singularity exec --no-home -B$ROOT \
        $DIR/apps/ragtag_2.1.0--pyhb7b1952_0.sif \
        ragtag.py "$@"

}

ragtag scaffold -u -t 48 -q 60 -o $DIR/kd/3-ragtag/ -i 0.75  \
    $DIR/0-ref/kirkii.fa $DIR/kd/canu.contigs.fasta
  #+end_src


** sup model

Using the 'Nanopore flip-flop R9.4 or R10.3' settings because of the reduced
error rate for the latest basecalling data.

#+begin_src sh :tangle kd.sup/run.sh
RAW=$(realpath $DIR/../0-raw/)
PATH=$DIR/apps/canu-2.2/bin:$PATH

canu \
    -p canu -d $DIR/kd.sup \
    genomeSize=590m \
    gridOptions="-Acotton_genomics" \
    useGrid=true \
    corMhapOptions='--threshold 0.8 --ordered-sketch-size 1000 --ordered-kmer-size 14' \
    correctedErrorRate=0.105 \
    -nanopore $RAW/*/sup.pass.fq.gz


#+end_src

| n      |               2365 |
| n:500  |               2365 |
| L50    |                 89 |
| min    |               1206 |
| N75    |             755687 |
| N50    |            1659488 |
| N25    |            3113363 |
| E-size |            2550534 |
| max    |            11.36e6 |
| sum    |            552.8e6 |
| name   | canu.contigs.fasta |

- Clean canu directory (see previous)


* Kokia kauaiensis S9

- 1st run
  #+begin_src sh :tangle kk/run.sh
RAW=$(realpath $DIR/../0-raw/)
PATH=$DIR/apps/canu-2.2/bin:$PATH

canu \
    -p canu -d $DIR/kk \
    genomeSize=590m \
    gridOptions="-Acotton_genomics" \
    useGrid=true \
    -nanopore $RAW/Kk/*/fastq_pass.fq.gz


#+end_src
| n      |                 16092 |
| n:500  |                 16092 |
| L50    |                  2530 |
| min    |                  1114 |
| N75    |                 27190 |
| N50    |                 56809 |
| N25    |                108701 |
| E-size |                 84051 |
| max    |                641467 |
| sum    |               536.3e6 |
| name   | kk/canu.contigs.fasta |


* Kokia cookei C69

- 1st run
  #+begin_src sh :tangle kc/run.sh
RAW=$(realpath $DIR/../0-raw/)
PATH=$DIR/apps/canu-2.2/bin:$PATH

canu \
    -p canu -d $DIR/kc \
    genomeSize=590m \
    gridOptions="-Acotton_genomics" \
    useGrid=true \
    -nanopore $RAW/Kc/*/fastq_pass.fq.gz


#+end_src
| n      |                  4992 |
| n:500  |                  4992 |
| L50    |                   414 |
| min    |                  1027 |
| N75    |                147619 |
| N50    |                342606 |
| N25    |                702028 |
| E-size |                561401 |
| max    |               4034082 |
| sum    |               555.8e6 |
| name   | kc/canu.contigs.fasta |

- hifiasm test
  #+begin_src sh
RAW=$(realpath $DIR/../0-raw/)
PATH=$DIR/apps/hifiasm.src/:$PATH

hifiasm -t 48 -o $DIR/kc/hifiasm.test $DIR/kc/canu.trimmedReads.fasta.gz
hifiasm -t 48 -l0 -o $DIR/kc/hifiasm.test.l0 $DIR/kc/canu.trimmedReads.fasta.gz

awk '/^S/{print ">"$2;print $3}' kc/hifiasm.test.bp.p_ctg.gfa \
    > kc/hifiasm.test.bp.p_ctg.fa
awk '/^S/{print ">"$2;print $3}' kc/hifiasm.test.l0.bp.p_ctg.gfa \
    > kc/hifiasm.test.l0.bp.p_ctg.fa

singularity exec -B $ROOT \
    /apps/singularity-3/abyss/abyss-v2.1.5-7-deb_cv1.sif \
    /usr/lib/abyss/abyss-fac kc/hifiasm.test{,.l0}.bp.p_ctg.fa

  #+end_src

| n      |         3184 |            4597 |
| n:500  |         3184 |            4597 |
| L50    |          333 |             429 |
| min    |         1332 |            1332 |
| N75    |       170099 |          123370 |
| N50    |       403680 |          309824 |
| N25    |       714084 |          575584 |
| E-size |       542454 |          417095 |
| max    |      2722741 |         2075424 |
| sum    |      471.4e6 |         487.2e6 |
| name   | hifiasm.test | hifiasm.test.l0 |

- miniasm

  #+begin_src sh
RAW=$(realpath $DIR/../0-raw/)

cat $RAW/Kc/*/fastq_pass.fq.gz > kc/reads.fq.gz
$DIR/apps/minimap2 -I30G -x ava-ont -t48 kc/reads.fq.gz kc/reads.fq.gz | gzip -1 > kc/reads.paf.gz
$DIR/apps/miniasm -f kc/reads.fq.gz kc/reads.paf.gz > kc/miniasm.gfa
  #+end_src

- shasta
  #+begin_src sh
RAW=$(realpath $DIR/../0-raw/)

zcat $RAW/Kc/*/fastq_pass.fq.gz > kc/reads.fq

$DIR/apps/shasta-0.9.0 \
    --input $DIR/kc/reads.fq \
    --threads 48 \
    --config Nanopore-Oct2021 \
    --assemblyDirectory $DIR/kc/shasta/

 $DIR/apps/shasta-0.9.0 \
    --input $DIR/kc/reads.fq \
    --threads 48 \
    --config Nanopore-Oct2021 \
    --Reads.minReadLength 5000 \
    --assemblyDirectory $DIR/kc/shasta.5k/

  #+end_src
| n                    |   10522 |    3212 |
| n:500                |   10453 |    3144 |
| L50                  |    2085 |     397 |
| min                  |     503 |     552 |
| N75                  |   39058 |  211770 |
| N50                  |   70150 |  397785 |
| N25                  |  116194 |  638523 |
| E-size               |   84740 |  471734 |
| max                  |  415102 | 2141538 |
| sum                  | 475.5e6 | 501.2e6 |
| min read length name |     10k |      5k |
